substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/sources/SourceCache.java

```diff
@@ -331,112 +310,64 @@ protected Path tryCacheFile(Path filePath, Class<?> clazz) {
+     * @param sourcePath A path for the file that may nee dot be cached
+     * @param targetPath A path for the target file in the cache
      */
-    protected Path checkCacheFile(Path filePath, Class<?> clazz) {
-        Path targetPath = cachedPath(filePath);
-        String moduleName = null;
-        if (clazz != null) {
-            /* Paths require the module name as prefix */
-            moduleName = clazz.getModule().getName();
-        }
-
-        if (moduleName != null) {
-            for (String specialRootModule : specialRootModules) {
-                if (moduleName.equals(specialRootModule)) {
-                    // handle this module specially as it has intermediate dirs
-                    for (Path srcRoot : specialSrcRoots.get(specialRootModule)) {
-                        String srcRootGroup = srcRoot.subpath(1, 2).toString().replace(".", filePath.getFileSystem().getSeparator());
-                        if (filePath.toString().startsWith(srcRootGroup)) {
-                            Path sourcePath = extendPath(srcRoot, filePath);
-                            try {
-                                if (tryCheckCacheFile(sourcePath, targetPath)) {
-                                    return filePath;
-                                }
-                            } catch (IOException e) {
-                                /* delete the target file as it is invalid */
-                                targetPath.toFile().delete();
-                                /* have another go at caching it */
-                                return tryCacheFile(filePath, clazz);
-                            }
-                        }
-                    }
-                    break;
-                }
-            }
-        }
+    void maybeCacheSource(Source source, Path sourcePath, Path targetPath) {
+        // see if the target file already exists
+        File cachedFile = targetPath.toFile();
+        boolean doCopy;
 
-        for (SourceRoot root : srcRoots) {
-            final Path scopedFilePath;
-            if (moduleName != null && root.isJDK) {
-                scopedFilePath = Paths.get(moduleName, filePath.toString());
-            } else {
-                scopedFilePath = filePath;
-            }
-            final Path sourcePath = extendPath(root.path, scopedFilePath);
+        if (cachedFile.exists()) {
             try {
-                if (tryCheckCacheFile(sourcePath, targetPath)) {
-                    return filePath;
-                }
+                // only copy if the size is different or the target is older than the source
+                FileTime sourceTime = Files.getLastModifiedTime(sourcePath);
+                FileTime destTime = Files.getLastModifiedTime(targetPath);
+                long sourceSize = Files.size(sourcePath);
+                long destSize = Files.size(targetPath);
+                doCopy = sourceSize != destSize || destTime.compareTo(sourceTime) < 0;
```

Comment:
What if the size is different but the target file is newer than the source? Shouldn't we through an error in that case or at least keep the cached one?

Reply:
Not necessarily. Firstly, this is all a bit of a heuristic. The way to properly ensure you get the right files is to nuke the cache.

A change of size is the most reliable indicator that we have a different jar but it is not fool proof. The time stamp check fixes some cases but it is not really that reliable. Imagine two different versions of a file Foo.java are in foo-1.0.jar and foo-1.1.jar but happen to have the same size. Consider what might happen if the user upgrades version from 1.0 to 1.1.

- If both versions are downloaded from Maven and were minted before the first cache operation took place then the cache copy of Foo.java will be newer than both of the jar versions (the copy does not preserve the original time stamps) so we will not recache -- bzzzt!
- If foo-1.1.jar has been released since we last cached then its Foo.java will be newer so we recache - ding!
- If foo-1.0 was built locally, Foo.java was cached and then a local foo-1.0 was built then we recache -- ding!
- If foo-1.1 was built locally and just happened to be built after the local foo-1.0 then we don't recache -- bzzzt!